lib/Sema/CSSimplify.cpp

```diff
@@ -3939,9 +3940,19 @@ ConstraintSystem::simplifyKeyPathConstraint(Type keyPathTy,
     break;
   }
   
+  // FIXME: Allow the type to be upcast if the type system has a concrete
+  // KeyPath type assigned to the expression already.
+  if (keyPathBGT) {
+    if (keyPathBGT->getDecl() == getASTContext().getKeyPathDecl())
+      kpDecl = getASTContext().getKeyPathDecl();
+    else if (keyPathBGT->getDecl() == getASTContext().getWritableKeyPathDecl()
+             && capability >= Writable)
+      kpDecl = getASTContext().getWritableKeyPathDecl();
+  }
```

Comment:
@DougGregor @rudkx It seems like this shouldn't be necessary, since even if the KeyPathExpr has a subclass type it could still satisfy a system with a conversion to the base class `KeyPath`, and it works without this code for conversions to `Any`/`AnyObject`. Any idea what I might be doing wrong here?